### PR TITLE
Use a single wallet batch for UpgradeKeyMetadata

### PR DESCRIPTION
**Pull request by jonasschnelli**
Opening wallets (the first time) after #14021 took on my end around 30 seconds due to the keymetadata migration (tested on regtest).

Using a single wallet batch reduces the required time for the migration down to <1s on my system for a default 2k keypool wallet.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -392,10 +393,11 @@ void CWallet::UpgradeKeyMetadata()
             // Write meta to wallet
             CPubKey pubkey;
             if (GetPubKey(meta_pair.first, pubkey)) {
-                WriteKeyMetadata(meta, pubkey, true);
+                batch.WriteKeyMetadata(meta, pubkey, true);
             }
         }
     }
+    batch.TxnCommit(); //commit before setting the flag
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: I used the explicit commit (rather then the RAII destructor) to avoid refolding the code.

**Comment by MeshCollider** [comment]
Tagged for 0.18 because I'd consider this a bugfix

**Comment by Sjors** [comment]
Concept ACK.

We should also consider doing that for `importmulti` (at least per descriptor).

**Comment by laanwj** [comment]
utACK and concept ACK 0bedcbafdfaa0e980ad91c0ba4b9039575017f0e

**Comment by MeshCollider** [comment]
utACK https://github.com/bitcoin/bitcoin/commit/0bedcbafdfaa0e980ad91c0ba4b9039575017f0e

@achow101 want to take a look?

**Review by promag — COMMENTED** [review]
Concept ACK.

On line 387, before throwing, shouldn't it commit the batch?

**Comment by MarcoFalke** [comment]
@promag Wouldn't it doe that, as the unique ptr gets destroyed (and flush on exit was set to `true` by default)?

**Comment by promag** [comment]
@MarcoFalke yap, it was clear. Maybe worth a comment.

@jonasschnelli is there a rationale for 1000?

**Comment by achow101** [comment]
utACK 0bedcbafdfaa0e980ad91c0ba4b9039575017f0e

**Comment by jonasschnelli** [comment]
> On line 387, before throwing, shouldn't it commit the batch?

As @MarcoFalke said, if the unique pointer gets out of the scope, it "commits" (flushes) to the database.

> @jonasschnelli is there a rationale for 1000?

Not really. I thought 1000 is a good compromise between cache memory requirements and performance. A new constant looked after an overkill. Initially I wanted to tie it to the keypool size constant but somehow I thought then that this has nothing to do with memory consumption.

**Comment by MarcoFalke** [comment]
1000 seems like a good upper bound, since the speedup flattens out after this:

upgradekeymetadata time

**Pull request merged**